common/src/main/java/com/zegelin/prometheus/exposition/FormattedByteChannel.java

```diff
@@ -1,32 +1,50 @@
 package com.zegelin.prometheus.exposition;
 
+import com.google.common.annotations.VisibleForTesting;
+
 import java.nio.ByteBuffer;
 import java.nio.channels.ReadableByteChannel;
 
 public class FormattedByteChannel implements ReadableByteChannel {
-    public static final int MIN_CHUNK_SIZE = 1024 * 1024;
-    public static final int MAX_CHUNK_SIZE = MIN_CHUNK_SIZE * 5;
+    public static final int DEFAULT_CHUNK_THRESHOLD = 1024 * 1024;
+    public static final int MAX_CHUNK_SIZE = DEFAULT_CHUNK_THRESHOLD * 5;
 
     private final FormattedExposition formattedExposition;
+    private final int chunkThreshold;
+
+    public FormattedByteChannel(final FormattedExposition formattedExposition) {
+        this(formattedExposition, DEFAULT_CHUNK_THRESHOLD);
+    }
 
-    public FormattedByteChannel(FormattedExposition formattedExposition) {
+    @VisibleForTesting
+    FormattedByteChannel(final FormattedExposition formattedExposition, final int chunkThreshold) {
         this.formattedExposition = formattedExposition;
+        this.chunkThreshold = chunkThreshold;
     }
 
     @Override
-    public int read(ByteBuffer dst) {
+    public int read(final ByteBuffer dst) {
         if (!isOpen()) {
             return -1;
         }
 
+        // Forcing the calling ChunkedNioStream to flush the buffer
+        if (hasBufferReachedChunkThreshold(dst)) {
+            return -1;
```

Comment:
The strategy used here is somewhat questionable.

Returning -1 while there's still data to read isn't exactly suggested by the `ReadableByteChannel` API. On the other hand, the ChunkedNioStream class, which is consuming this byte channel in Netty, seem to expect this behavior deliberately. I've verified this to work on both Netty 4.0 and 4.1.

This was the most elegant solution I could come up with that will avoid another staging buffer, and at the same time stay compatible with both Netty 4.0 and 4.1.